content/article/hotel-hotspot-hijinks.md

```diff
@@ -0,0 +1,300 @@
+0  HTTP::Response=HASH(0x55a95f5048c0)
+<snip>
+lots of details; you really don't want to see this
+</snip>
+```
+
+That's what we like to see!  We're at least getting stuff back now.  Having
+a look at the page's title, we get:
+
+```
+  DB<5> x $mech->title();
+0  'myadvise hotspot > login'
+```
+
+Yup, that's a login page.  Dumping the page's content with
+
+```
+  DB<5> x $mech->content();
+<snip>
+lots of HTML content
+</snip>
+```
+
+we get to see what we've got to play with.  The main things to note about
+the content (which I'm not showing because it's too much detail and I want
+to protect the innocent) are:
+
+  - we have a form called `login`
+
+```html
+<form name="login" action="http://login.hotspot.lan/login" method="post">
+```
+
+  - we have a username field with the name `username`
+
+```html
+<input style="width: 80px" name="username" type="text" value=""/>
+```
+
+  - and we have a password field with the name `password`
+
+```html
+<input style="width: 80px" name="password" type="password"/>
+```
+
+This gives us enough information to be able to submit the form using the
+relevant login data.
+
+Aside: interestingly enough, the fields are in English even though the site
+is a German one.  I guess standardising the fields on English can be useful
+when programming.
+
+To submit the form, we use `WWW::Mechanize`'s
+[`submit_form()`](https://metacpan.org/pod/WWW::Mechanize#$mech-%3Esubmit_form(-...-))
+method (the call to which I've formatted nicely here to make things easier
+to read):
+
+```perl
+$mech->submit_form(
+    form_name => 'login',
+    fields => {
+        username => 'username-for-room',
+        password => 'password-for-room',
+    }
+);
+```
+
+We can check if the form submission was successful by asking the
+[`HTTP::Response`](https://metacpan.org/pod/HTTP::Response) if things went
+well:
+
+```
+  DB<7> x $mech->res->is_success;
+0  1
+```
+
+Looking good so far.  Let's see if `ping` works as expected
+
+```
+$ ping www.heise.de
+PING www.heise.de (193.99.144.85) 56(84) bytes of data.
+64 bytes from www.heise.de (193.99.144.85): icmp_seq=1 ttl=247 time=20.6 ms
+64 bytes from www.heise.de (193.99.144.85): icmp_seq=2 ttl=247 time=13.9 ms
+^C
+--- www.heise.de ping statistics ---
+2 packets transmitted, 2 received, 0% packet loss, time 1000ms
+rtt min/avg/max/mdev = 13.941/17.281/20.621/3.340 ms
+```
+
+Yes!  In other words: we're in!  [That was
+easy](https://www.youtube.com/watch?v=kCE0v-SK5_c). &#x263A;
+
+Putting everything together (and cleaning up the code a bit), I ended up
+with this:
+
+```perl
+use strict;
+use warnings;
+
+use WWW::Mechanize;
+
+my $mech = WWW::Mechanize->new;
+$mech->get('http://google.com');
+
+# check that we have the login page and not Google or something else
+# i.e. we're not logged in
+if ($mech->title() =~ 'login') {
+    $mech->submit_form(
+        form_name => 'login',
+        fields => {
+            username => 'whg4@haus-reimann',
+            password => 'HRAltenau4'
```

Comment:
First off, thanks very much! Someone pointed out that this is probably enough information to deduce the name of the hotel (and possibly the room in question). I think obscuring that would not make the article any less useful?

Reply:
Bother!  I obviously forgot to change that after copy-pasting the "production" code into the article. Damn!  Thanks heaps for pointing that out!

Reply:
Ok, I've fixed the offending lines.  I feel really daft now for having added that.  It certainly wasn't my intention :-/

Reply:
I once left hard-coded credentials in a script that was included in a blog post. A very kind soul emailed me to point that out before any harm was done.

Reply:
Thanks for helping me out on that one; I appreciate it! :-)